katdal/applycal.py

```diff
@@ -198,25 +198,38 @@ def calc_bandpass_correction(sensor, index, data_freqs, cal_freqs):
     return CategoricalData(corrections, sensor.events)
 
 
-def calc_gain_correction(sensor, index):
+def calc_gain_correction(sensor, index, targets=None):
     """Calculate correction sensor from gain calibration solution sensor.
 
     Given the gain calibration solution `sensor`, this extracts the time
     series of gains for the input specified by `index` (in the form (pol, ant))
     and interpolates them over time to get the corresponding complex correction
-    terms.
+    terms. The optional `targets` parameter is a `CategoricalData` or array of
+    target indices, i.e. a sensor indicating the target associated with each
+    dump. If provided, interpolate solutions derived from one target only at
+    dumps associated with that target, which is what you want for
+    self-calibration solutions.
 
     Invalid solutions (NaNs) are replaced by linear interpolations over time
     (separately for magnitude and phase), as long as some dumps have valid
-    solutions.
+    solutions on the appropriate target.
     """
     dumps = np.arange(sensor.events[-1])
     events = sensor.events[:-1]
-    gains = np.array([value[index] for segment, value in sensor.segments()])
-    valid = np.isfinite(gains)
-    if not valid.any():
-        return CategoricalData([INVALID_GAIN], [0, len(dumps)])
-    smooth_gains = complex_interp(dumps, events[valid], gains[valid])
+    gains = np.array([value[(Ellipsis,) + index]
+                      for segment, value in sensor.segments()])
+    # Let the gains be shaped either (n_cal_freqs, n_events) or (1, n_events)
+    gains = np.atleast_2d(gains.T)
+    # Assume all dumps have the same target by default, i.e. interpolate freely
+    if targets is None:
+        targets = CategoricalData([0], [0, len(dumps)])
+    smooth_gains = np.empty((len(dumps), gains.shape[0]), dtype=gains.dtype)
+    for chan, gains_per_chan in enumerate(gains):
+        for target in set(targets):
+            on_target = (targets == target)
+            valid = np.isfinite(gains_per_chan) & on_target[events]
+            smooth_gains[on_target, chan] = INVALID_GAIN if not valid.any() else \
+                complex_interp(dumps[on_target], events[valid], gains_per_chan[valid])
```

Comment:
It's actually O(cal_n_chans x n_dumps), since the dumps are selected per target via the `on_target` mask and each dump is therefore fitted only once per channel. And yes, cal_n_chans is the number of contim subbands. I've added a comment as a reminder.

Reply:
The complex_interp is O(cal_n_chans × dumps), but the computation of and indexing with `valid` is O(cal_n_chans × dumps × n_targets) because `valid` has size `n_dumps` even if it's mostly False. But the `complex_interp` is likely to dominate in practice, which is why I'm not too worried for now.

Reply:
Right, my thinking too.